geometry/sat_collision.cpp

```diff
@@ -0,0 +1,200 @@
+    Vec2 closestVertex = otherShape.GetClosestVertex(m_Centre);
+    Vec2 axis = closestVertex - m_Centre;
+    axis.Normalize();
+    return OverlapOnAxis(axis, *this, otherShape);
+}
+Vec2 Circle::GetClosestVertex(Vec2 point) const {
+    // Not technically a vertex but the (only) closest stored point
+    return m_Centre;
+}
+void Circle::Project(Vec2 axis, double& minRes, double& maxRes) const {
+    axis.Normalize();
+    minRes = (m_Centre - (axis * m_Radius)).ScalarProjection(axis);
+    maxRes = (m_Centre + (axis * m_Radius)).ScalarProjection(axis);
+}
+/******************************************************************************/
+/********************************* SAT ****************************************/
+bool DoOverlap(double min1, double max1, double min2, double max2) {
+    return !((max2 < min1) || (max1 < min2));
+}
+
+bool OverlapOnAxis(const Vec2& axis, const Shape& shape1,
+              const Shape& shape2) {
+    double min1, max1, min2, max2;
+    shape1.Project(axis, min1, max1);
+    shape2.Project(axis, min2, max2);
+    return DoOverlap(min1, max1, min2, max2);
+}
+
+bool sat(const Shape& shape1, const Shape& shape2) {
+    return shape1.SAT(shape2) && shape2.SAT(shape1);
+}
+/******************************************************************************/
+}}
+/********************************* TEST ***************************************/
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    geometry::collision::Circle c1({2.0, 3.0}, 1.0);
+    geometry::collision::Circle c2({1.5374,4.79596}, 1.0);
+    geometry::collision::Circle c3({5.5, 2.0}, 2.0);
+
+    geometry::collision::Polygon p1({{1.0, 1.0}, {3.0, 3.0}, {3.0, 1.0}});
+    geometry::collision::Polygon p2({{2.0, 2.5}, {4.0, 2.7}, {3.0, 4.0}});
+    geometry::collision::Polygon p3({{2.7, 4.5}, {4.6, 4.2}, {3.8, 3.4}});
+
+    // poly-circle collision
+    assert(geometry::collision::sat(p1, c1));
+    assert(!geometry::collision::sat(c2, p1));
+    assert(geometry::collision::sat(p2, c3));
+
+    // poly-poly collision
+    assert(geometry::collision::sat(p1, p2));
+    assert(!geometry::collision::sat(p1, p3));
+    assert(!geometry::collision::sat(p2, p3));
+
+    // circle-circle collision
+    assert(geometry::collision::sat(c1, c2));
+    assert(!geometry::collision::sat(c1, c3));
+    assert(!geometry::collision::sat(c2, c3));
+
+    // A shape should always collide with itself
+    assert(geometry::collision::sat(c1, c1));
+    assert(geometry::collision::sat(p1, p1));
+
+    std::cout << "All tests passed!" << std::endl;
+}
+/******************************************************************************/
+/*********************************** MAIN *************************************/
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();
+    return 0;
+}
+/******************************************************************************/
```

Comment:
Missing an endline. Without this endline, there could be issues with some IDEs and configurations.

```suggestion
/******************************************************************************/

```


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Inefficient